internal/envoy/v3/listener.go

```diff
@@ -308,34 +317,56 @@ func (b *httpConnectionManagerBuilder) DefaultFilters() *httpConnectionManagerBu
 	// Add a default set of ordered http filters.
 	// The names are not required to match anything and are
 	// identified by the TypeURL of each filter.
-	b.filters = append(b.filters,
-		&envoy_filter_network_http_connection_manager_v3.HttpFilter{
-			Name: CompressorFilterName,
-			ConfigType: &envoy_filter_network_http_connection_manager_v3.HttpFilter_TypedConfig{
-				TypedConfig: protobuf.MustMarshalAny(&envoy_filter_http_compressor_v3.Compressor{
-					CompressorLibrary: &envoy_config_core_v3.TypedExtensionConfig{
-						Name: "gzip",
-						TypedConfig: protobuf.MustMarshalAny(
-							&envoy_compression_gzip_compressor_v3.Gzip{},
-						),
-					},
-					ResponseDirectionConfig: &envoy_filter_http_compressor_v3.Compressor_ResponseDirectionConfig{
-						CommonConfig: &envoy_filter_http_compressor_v3.Compressor_CommonDirectionConfig{
-							ContentType: []string{
-								// Default content-types https://github.com/envoyproxy/envoy/blob/e74999dbdb12aa4d6b7a5d62d51731ea86bf72be/source/extensions/filters/http/compressor/compressor_filter.cc#L35-L38
-								"text/html", "text/plain", "text/css", "application/javascript", "application/x-javascript",
-								"text/javascript", "text/x-javascript", "text/ecmascript", "text/js", "text/jscript",
-								"text/x-js", "application/ecmascript", "application/x-json", "application/xml",
-								"application/json", "image/svg+xml", "text/xml", "application/xhtml+xml",
-								// Additional content-types for grpc-web https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-WEB.md#protocol-differences-vs-grpc-over-http2
-								"application/grpc-web", "application/grpc-web+proto", "application/grpc-web+json", "application/grpc-web+thrift",
-								"application/grpc-web-text", "application/grpc-web-text+proto", "application/grpc-web-text+thrift",
+	var compressor proto.Message = &envoy_compression_gzip_compressor_v3.Gzip{}
+	compressorName := string(contour_v1alpha1.GzipCompression)
+	if b.compression != nil {
```

Comment:
It's true that the way the code is at present means the order does matter. At the same time the intent of the PR is to allow a way of changing the compression algorithm in the default filter chain (as opposed to something more generic), so it kind of makes sense that `DefaultFilters()` no longer sets up a fixed set.  Maybe `Compression()` is a bad name for that method, perhaps it should be something like `OverrideDefaultCompression()`, and documented more clearly to explain the behaviour including how it needs to be called before `DefaultFilters()` if it is to do anything.

I've been trying to avoid changing the code in any way that would risk changing behaviour for users who don't opt in to the new method. At a basic level this certainly means we can't change the signatures say of  `HTTPConnectionManagerBuilder()` or `DefaultFilters()` itself, but I would also want to avoid changing the behaviour of `DefaultFilters()` in the absence of using the new compression method. So in thinking about these options:

About 1. Remove compressor from the default set entirely. 
I wouldn't be keen to do this as the behaviour of adding compression is effectively already part of the contract of what `DefaultFilters()` does. Changing this could change the behaviour of downstream users' code.

Option 3 is similar, in that you're changing the order of filters (currently default filters put compression filter first). Less dramatic than option 1 but still a change in behaviour.

Option 2 might be possible - I guess `Get()` could scan the filter chain looking for one named  "envoy.filters.http.compressor" (value of `CompressorFilterName` as set in `DefaultFilters()` at present) to update.  Would that be good enough? Can it update the first such filter it finds, on the assumption that there will be only one?

I'd like to be sure about any changes before I make them here, but my feeling is that perhaps no change is needed -  other than renaming the `Compression()` func and providing better documentation. What do you think?